aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -2989,98 +3017,165 @@ std::tuple<Tensor,Tensor> eig(const Tensor& self, bool eigenvectors) {
+  const auto Vh_ready = !compute_uv
+                            || (!use_cusolver && Vh.mT().is_contiguous())
+                            || (use_cusolver && Vh.is_contiguous());
+  const auto Vh_ = borrow_else_clone(Vh_ready, Vh, Vh, /*C-contig*/use_cusolver);
+
+  // We need the const_cast because some of the cusolver routines compute V rather than Vh
+  // so we need to compute the output.
+  svd_stub(A.device().type(),
+           A,
+           full_matrices,
+           compute_uv,
+           *U_, *S_, const_cast<Tensor&>(*Vh_), info);
+
+  if (!U_ready) {
+    U.copy_(*U_);
+  }
+  if (!S_ready) {
+    S.copy_(*S_);
+  }
+  if (!Vh_ready) {
+    Vh.copy_(*Vh_);
+  }
+
+  // TODO This should be removed, and the code checking for convergence should be lifted
+  // from svd_cusolver to this function. As such, we should make sure that this function
+  // never returns an error.
+  if (A.dim() > 2) {
+    batchCheckErrors(info, "linalg.svd");
+  } else {
+    singleCheckErrors(info.item<int64_t>(), "linalg.svd");
+  }
+}
+
+TORCH_IMPL_FUNC(linalg_svd_out)(const Tensor& A,
+                                bool full_matrices,
+                                const Tensor & U,
+                                const Tensor & S,
+                                const Tensor & Vh) {
+  // This function does not have an _ex variant as we always check errors inside
+  // to assure the convergence of the algorithm anyway. See
+  // https://github.com/pytorch/pytorch/issues/28293
+  // https://github.com/pytorch/pytorch/issues/64237
+  linalg_svd_and_svdvals(A,
+                         full_matrices,
+                         /*compute_uv=*/true,
+                         const_cast<Tensor&>(U),
+                         const_cast<Tensor&>(S),
+                         const_cast<Tensor&>(Vh));
+}
+// Can't make it structured as its a composite kernel
+// See note in linalg_svd for why this function does not have an _ex variant
+Tensor& linalg_svdvals_out(const Tensor& A, Tensor & S) {
+  checkLinalgCompatibleDtype(
+      "torch.linalg.svdvals", S.scalar_type(), toValueType(A.scalar_type()));
 
-std::tuple<Tensor, Tensor, Tensor> _svd_helper_cpu(const Tensor& self, bool some, bool compute_uv) {
-  std::vector<int64_t> infos(batchCount(self), 0);
+  // Resize S
+  auto sizes = A.sizes().vec();
+  sizes.pop_back();
+  sizes.end()[-1] = std::min(A.size(-2), A.size(-1));
+  at::native::resize_output(S, sizes);
 
-  char jobz = compute_uv ? (some ? 'S' : 'A') : 'N';
+  Tensor U, Vh;  // Dummies
+  linalg_svd_and_svdvals(A, /*full_matrices=*/false, /*compute_uv=*/false, U, S, Vh);
 
-  Tensor U_working_copy, S_working_copy, VT_working_copy;
-  std::tie(U_working_copy, S_working_copy, VT_working_copy) = _create_U_S_VT(self, some, compute_uv);
+  return S;
+}
 
-  auto self_working_copy = cloneBatchedColumnMajor(self);
+Tensor linalg_svdvals(const Tensor& A) {
+  return std::get<1>(at::linalg_svd(A, /*full_matrices=*/false));
```

Comment:
Lol, just saw this. This is pretty much the fix I had done
![image](https://user-images.githubusercontent.com/3291265/149334066-fb98957d-4ab0-4ff4-a9af-58535ec0843e.png)
Note the pedantic note of returning `S` rather than what `svdvals_out` returns, as to avoid a tensor copy (and a ref bump), as `svdvals_out` returns a `Tensor&` :P

Reply:
Maybe it's better to use `at::linalg_svdvals_outf` instead of `at::native::linalg_svdvals_out`

Reply:
It does not work either, because it wouldn't be a composite function anymore (the composite tests complain)

Reply:
(but yeha, we both wrote the order of the parameters swapped above)